### PR TITLE
inbound: Handle direct connections with a dedicated stack

### PR DESCRIPTION
**Pull request by olix0r**
Currently, the inbound proxy handles direct connections--connections
where the original destination is the inbound proxy port--in two places.
TCP forwarding is handled distinctly from HTTP forwarding (for
gateways).

This change modifies the inbound proxy to dispatch these connections to
a dedicated `direct` stack that does its own protocol detection and
routing.

This sets up to (1) modify the server to configure ALPN on only these
connections and (2) support transport-header-based gateway routing.

### CODE REVIEW

File: linkerd/app/inbound/src/lib.rs [hunk 1]
```diff
@@ -127,32 +123,88 @@ impl Config {
...
+                .push_map_target(
+                    |(v, (h, mut t)): (
+                        Option<http::Version>,
+                        (Option<TransportHeader>, TcpAccept),
+                    )| {
+                        if let Some(h) = h {
+                            t.target_addr = (t.target_addr.ip(), h.port).into();
+                        }
+                        (v, t)
+                    },
```
Comment (hawkw) [inline linkerd/app/inbound/src/lib.rs:175]: yeesh.

(i imagine this is going to change soon though, based on the comment?)

File: linkerd/app/inbound/src/prevent_loop.rs [hunk 2]
```diff
@@ -1,5 +1,10 @@
+use crate::TcpEndpoint;
 //use futures::future;
```
Comment (hawkw) [inline linkerd/app/inbound/src/prevent_loop.rs:2]: nit: looks like this was already there, but should we delete it?

File: linkerd/app/inbound/src/require_identity.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+impl Predicate<TcpAccept> for RequireIdentityForDirect {
+    type Request = TcpAccept;
+
+    fn check(&mut self, meta: TcpAccept) -> Result<TcpAccept, Error> {
+        tracing::debug!(client.id = ?meta.client_id);
+        match meta.client_id {
+            Conditional::Some(Some(_)) => {}
+            _ => return Err(IdentityRequired(()).into()),
+        }
+        Ok(meta)
```
Comment (hawkw) [inline linkerd/app/inbound/src/require_identity.rs:32]: nit/tioli: isn't this just 
```suggestion
        match meta.client_id {
            Conditional::Some(Some(_)) => Ok(meta),
            _ =>Err(IdentityRequired(()).into(),
        }
```
or does the borrow checker not like that?

File: linkerd/app/inbound/src/require_identity.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+/// A connection policy that fails connections that don't have a client identity
+/// if they target one of the configured local ports.
+#[derive(Clone, Debug)]
+pub struct RequireIdentityForPorts {
+    ports: Arc<IndexSet<u16>>,
+}
+
+#[derive(Debug)]
+pub struct IdentityRequired(());
```
Comment (hawkw) [inline linkerd/app/inbound/src/require_identity.rs:19]: now that we return this in multiple places, should we add a scope to it? or will it be clear from the logs where it came from?

**Review by hawkw — APPROVED** [review]
this looks good to me! i had a few minor nits.

File: linkerd/app/inbound/src/require_identity.rs (hunk 4 above)
Comment (olix0r) [reply linkerd/app/inbound/src/require_identity.rs:19]: it's effectively the same error in both cases. It's as if 4143 is in the required ports list (which i believe it is, anyway).

**Pull request merged**